[LV] Add scalar load/stores to VPReplicateRecipe::computeCost

david-arm opened the pull request:
Avoid calling getLegacyCost for single scalar loads and stores where the cost is trivial to calculate.

david-arm commented:
Gentle ping. :)

fhahn commented on llvm/lib/Transforms/Vectorize/LoopVectorize.cpp:
```diff
@@ -5205,6 +5205,8 @@ LoopVectorizationCostModel::getConsecutiveMemOpCost(Instruction *I,
     Cost += TTI.getMaskedMemoryOpCost(I->getOpcode(), VectorTy, Alignment, AS,
                                       CostKind);
   } else {
+    // TODO: Is this right? Operand 0 has a different meaning for loads and
+    // stores.
```
It is indeed confusing and the argument isn't documented, but it looks like in RISCV TTI and X86 TTI (https://github.com/llvm/llvm-project/blob/main/llvm/lib/Target/X86/X86TargetTransformInfo.cpp#L5241) the argument is used as it is passed here: stored value for stores, address for loads. AArch64 and ARM TTIs don't use it.

So it seems it is used as expected, even though the documentation of the interface could be improved

david-arm replied:
Yeah, sorry just got caught up with other things. I guess I wasn't really sure whether you wanted me to change the documentation in this PR or were just pointing it out for information purpose?

fhahn replied:
Probably best to remove the TOOD here and clarify the TTI separately?

david-arm committed:
[LV] Add scalar load/stores to VPReplicateRecipe::computeCost

Avoid calling getLegacyCost for single scalar loads and stores
where the cost is trivial to calculate.

fhahn reviewed (approved):
LGTM, thanks
